source/opt/dominator_tree.h

```diff
@@ -0,0 +1,146 @@
+
+  ir::BasicBlock* BB;
+  DominatorTreeNode* Parent;
+  std::vector<DominatorTreeNode*> Children;
+
+  // These indexes are used to compare two given nodes. A node is a child or
+  // grandchild of another node if its preorder index is greater than the
+  // first nodes preorder index AND if its postorder index is less than the
+  // first nodes postorder index.
+  int DepthFirstPreOrderIndex;
+  int DepthFirstPostOrderIndex;
+
+  // Returns the ID of the basic block. We need this method to be able to use
+  // this struct in the cfa depth first traversal function.
+  uint32_t id() const;
+};
+
+// A class representing a tree of BasicBlocks in a given function, where each
+// node is dominated by its parent.
+class DominatorTree {
+ public:
+  using DominatorTreeNodeList = std::vector<DominatorTreeNode*>;
+  using iterator = DominatorTreeNodeList::iterator;
+  using const_iterator = DominatorTreeNodeList::const_iterator;
+
+  DominatorTree() : PostDominator(false) {}
+  DominatorTree(bool post) : PostDominator(post) {}
+
+  iterator begin() { return Roots.begin(); }
+  iterator end() { return Roots.end(); }
+  const_iterator begin() const { return cbegin(); }
+  const_iterator end() const { return cend(); }
+  const_iterator cbegin() const { return Roots.begin(); }
+  const_iterator cend() const { return Roots.end(); }
+
+  // Get the unique root of the tree.
+  // It is guaranteed to work on a dominator tree post-dominator might have a
+  // list.
+  DominatorTreeNode* GetRoot() {
+    assert(Roots.size() == 1);
+    return *begin();
+  }
+
+  const DominatorTreeNode* GetRoot() const {
+    assert(Roots.size() == 1);
+    return *begin();
+  }
+
+  // Dumps the tree in the graphvis dot format into the stream.
+  void DumpTreeAsDot(std::ostream& OutStream) const;
+
+  // Build the (post-)dominator tree for the function |F|
+  // Any existing data will be overwritten
+  void InitializeTree(const ir::Function* F);
+
+  // Check if BasicBlock B is a dominator of BasicBlock A.
+  bool Dominates(const ir::BasicBlock* A, const ir::BasicBlock* B) const;
+
+  // Check if BasicBlock B is a dominator of BasicBlock A. This function uses
+  // the IDs of A and B.
+  bool Dominates(uint32_t A, uint32_t B) const;
+
+  // Check if BasicBlock A strictly dominates B
+  bool StrictlyDominates(const ir::BasicBlock* A,
+                         const ir::BasicBlock* B) const;
+
+  bool StrictlyDominates(uint32_t A, uint32_t B) const;
+
+  // Returns the immediate dominator of basicblock A.
+  ir::BasicBlock* ImmediateDominator(const ir::BasicBlock* A) const;
+
+  // Returns the immediate dominator of basicblock A.
+  ir::BasicBlock* ImmediateDominator(uint32_t A) const;
+
+  // Returns true if BasicBlock A is reachable by this tree. A node would be
+  // unreachable if it cannot be reached by traversal from the start node or for
+  // a postdominator tree, cannot be reached from the exit nodes.
+  bool Reachable(const ir::BasicBlock* A) const;
```

Comment:
Cool.  I think we can use this in CFGCleanup instead of the ad-hoc reachability analysis it does now (#1026)